runtime/android/core/src/org/xwalk/core/XWalkActivity.java

```diff
@@ -5,16 +5,720 @@
+            Log.d("XWalkActivity", "Download progress: " + progress[0] + "/" + progress[1]);
+            mProgressDialog.setProgress(progress[0]);
+            mProgressDialog.setMax(progress[1]);
+            mProgressDialog.setIndeterminate(false);
+        }
+
+        @Override
+        protected void onCancelled(Integer result) {
+            Log.d("XWalkActivity", "Download cancelled");
+            mXWalkActivity.finish();
+        }
+
+        @Override
+        protected void onPostExecute(Integer result) {
+            mProgressDialog.dismiss();
+
+            if (result == DownloadManager.STATUS_SUCCESSFUL) {
+                Log.d("XWalkActivity", "Download finished");
+                Uri uri = mDownloadManager.getUriForDownloadedFile(mDownloadId);
+                mXWalkActivity.installXWalkLibrary(uri);
+            } else {
+                String errMsg = XWalkMixedResources.DOWNLOAD_FAILED_MESSAGE;
+                if (result == DownloadManager.STATUS_FAILED) {
+                    Query query = new Query().setFilterById(mDownloadId);
+                    Cursor cursor = mDownloadManager.query(query);
+                    if (cursor != null && cursor.moveToFirst()) {
+                        int reasonIdx = cursor.getColumnIndex(DownloadManager.COLUMN_REASON);
+                        int reason = cursor.getInt(reasonIdx);
+                        if (reason == DownloadManager.ERROR_DEVICE_NOT_FOUND) {
+                            errMsg = XWalkMixedResources.DOWNLOAD_FAILED_DEVICE_NOT_FOUND ;
+                        } else if (reason == DownloadManager.ERROR_INSUFFICIENT_SPACE) {
+                            errMsg = XWalkMixedResources.DOWNLOAD_FAILED_INSUFFICIENT_SPACE ;
+                        }
+                    }
+                } else if (result == DownloadManager.STATUS_PAUSED) {
+                } else if (result == DownloadManager.STATUS_RUNNING) {
+                    errMsg = XWalkMixedResources.DOWNLOAD_FAILED_TIME_OUT;
+                } else {
+                    assert(false);
+                }
+
+                Log.d("XWalkActivity", "Download failed, " + errMsg);
+                new DownloadFailedDialog(mXWalkActivity, errMsg).show();
+            }
+        }
+    }
+    
+    private static class DecompressTask extends AsyncTask<Void, Integer, Void> {
+        XWalkActivity mXWalkActivity;
+        AlertDialog mDialog;
+
+        DecompressTask(XWalkActivity activity, AlertDialog dialog) {
+            super();
+            mXWalkActivity = activity;
+            mDialog = dialog;
+        }
+
+        @Override
+        protected Void doInBackground(Void... params) {
+            try {
+                XWalkCoreWrapper.decompressXWalkLibrary();
+                // TODO: use publishProgress to update percentage.
+            } catch (Exception e) {
+                Log.w("XWalkActivity", "decompress library failed: " + e.getMessage());
+            }
+
+            return null;
```

Comment:
doInBackground should pass the result of decompressXWalkLibrary to onPostExecute by return statement